src/sqlancer/clickhouse/gen/ClickHouseExpressionGenerator.java

```diff
@@ -53,48 +53,128 @@ private enum Expression {
-        if (columns.isEmpty() || depth > 0 && Randomly.getBoolean()) {
+    public ClickHouseExpression generateExpressionWithColumns(List<ClickHouseColumnReference> columns, int complexity) {
+        if (columns.isEmpty() || complexity <= 2 && Randomly.getBooleanWithRatherLowProbability()) {
             return generateConstant(null);
         }
 
-        if (depth >= 2 || Randomly.getBoolean()) {
+        if (complexity <= 2 || Randomly.getBooleanWithRatherLowProbability()) {
             return columns.get((int) Randomly.getNotCachedInteger(0, columns.size() - 1));
         }
 
         ColumnLike expr = Randomly.fromOptions(ColumnLike.values());
         switch (expr) {
         case UNARY_PREFIX:
-            return new ClickHouseUnaryPrefixOperation(generateExpressionWithColumns(columns, depth + 1),
+            return new ClickHouseUnaryPrefixOperation(generateExpressionWithColumns(columns, complexity - 1),
                     ClickHouseUnaryPrefixOperator.MINUS);
         case BINARY_ARITHMETIC:
-            return new ClickHouseBinaryArithmeticOperation(generateExpressionWithColumns(columns, depth + 1),
-                    generateExpressionWithColumns(columns, depth + 1),
+            return new ClickHouseBinaryArithmeticOperation(generateExpressionWithColumns(columns, complexity - 1),
+                    generateExpressionWithColumns(columns, complexity - 1),
                     ClickHouseBinaryArithmeticOperation.ClickHouseBinaryArithmeticOperator.getRandom());
         case UNARY_FUNCTION:
-            return new ClickHouseUnaryFunctionOperation(generateExpressionWithColumns(columns, depth + 1),
+            return new ClickHouseUnaryFunctionOperation(generateExpressionWithColumns(columns, complexity - 1),
                     ClickHouseUnaryFunctionOperation.ClickHouseUnaryFunctionOperator.getRandom());
         case BINARY_FUNCTION:
-            return new ClickHouseBinaryFunctionOperation(generateExpressionWithColumns(columns, depth + 1),
-                    generateExpressionWithColumns(columns, depth + 1),
+            return new ClickHouseBinaryFunctionOperation(generateExpressionWithColumns(columns, complexity - 1),
+                    generateExpressionWithColumns(columns, complexity - 1),
                     ClickHouseBinaryFunctionOperation.ClickHouseBinaryFunctionOperator.getRandom());
         default:
             throw new AssertionError(expr);
         }
     }
 
+    public ClickHouseExpression generateAggregateExpressionWithColumns(List<ClickHouseColumnReference> columns,
+            int complexity) {
+        if (Randomly.getBooleanWithRatherLowProbability()) {
+            return new ClickHouseAggregate(generateExpressionWithColumns(columns, complexity - 1),
+                    ClickHouseAggregate.ClickHouseAggregateFunction.getRandom());
+        }
+        if (columns.isEmpty() || complexity <= 2 && Randomly.getBooleanWithRatherLowProbability()) {
```

Comment:
Instead of the raw integers `2`, perhaps it might improve maintainability to introduce a constant?

Reply:
I really don't have any clue which numbers should I use there.
There is this minimum depth where I may generate some trivial leaves with constants or columns. There is depth itself and I saw strange things that will smaller depth expressions were incorrect with higher probability. Maybe I should treat other places where I put constants where I generate different types of expressions(where/order by/join on). Actually I did something like that while testing the result.